docs/_includes/Statements.md

```diff
@@ -239,6 +240,242 @@ Note that the form
+```dafny
+{% include Example-Fail1.dfy %}
+```
+
+A commonly used alternative that carries some value information is something like this generic type:
+```dafny
+{% include Example-Fail2.dfy %}
+```
+
+A failure-compatible type with an `Extract` method is called _value-carrying_.
+
+
+### Simple status return with no other outputs
+
+The simplest use of this failure-return style of programming is to have a method call that just returns a non-value-carrying `Status` value:
+```dafny
+method Callee(i: int) returns (r: Status)
+{
+  if i < 0 { return Failure("negative"); }
+  return Success;
+}
+
+method Caller(i: int) returns (rr: Status)
+{
+  :- Callee(i);
+  ...
+}
+```
+
+Note that there is no LHS to the `:-` statement. 
+If `Callee` returns `Failure`, then the caller immediately returns, 
+not executing any statements following the call of `Callee`. 
+The value returned by `Caller` (the value of `rr` in the code above) is the result of `PropagateFailure` applied to the value returned by `Callee`, which is often just the same value. 
+If `Callee` does not return `Failure` (that is, returns a value for which `IsFailure()` is `false`) 
+then that return value is forgotten and execution proceeds normally with the statements following the call of `callee` in the body of `Caller`.
+
+### Status return with additional outputs
+
+The example in the previous subsection affects the program only through side effects or the status return itself. 
+It may well be convenient to have additional out-parameters, as is allowed for `:=` updates;
+these out-parameters behave just as for `:=`.
+Here is an example:
+
+```dafny
+method Callee(i: int) returns (r: Status, v: int, w: int)
+{
+  if i < 0 { return Failure("negative"); }
+  return Success(i), i+i, i*i;
+}
+
+method Caller(i: int) returns (rr: Status, k: int)
+{
+  var j: int;
+  j, k :- Callee(i);
+  k := k + k;
+}
+```
+
+Here `Callee` has two outputs in addition to the `Status` output. 
+The LHS of the `:-` statement accordingly has two l-values to receive those outputs. 
+Those outputs may be any sort of l-values;
+here they are a local variable and an out-parameter of the caller.
+Those outputs are assigned in the `:-` call regardless of the `Status` value:
+
+   * If `Callee` returns a failure value as its first output, then the other outputs are assigned, the _caller's_ first out-parameter (here `rr`) is assigned the value of `PropagateFailure`, and the caller returns.
+   * If `Callee` returns a non-failure value as its first output, then the other outputs are assigned and the 
+caller continues execution as normal.
+
+
+### Failure-returns with additional data
+
+The failure-compatible return value can carry additional data as shown in the `Outcome<T>` example above. 
+In this case there is a (first) LHS l-value to receive this additional data.
+
+```dafny
+method Callee(i: int) returns (r: Outcome<nat>, v: int)
+{
+  if i < 0 { return Failure("negative"); }
```

Comment:
Illegal `return` statement, see previous comments. In this case, the correction should be

    return Failure("negative"), -2;

since the text below says `v` returns as `-2` in the failure case.

Reply:
Reviewed and fixed all the return statements, including in desugarings